File: leetcode/medium/1339-max-prod-splitted/two-passes.py
"""
  @author Vortexx2
  Problem 1339 - Max Prod of Splitted BT

  Runtime - 400 ms
  Memory Usage - 69 MB
"""
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def maxProduct(self, root: Optional[TreeNode]) -> int:
    self.res = 0
    self.total = 0
    self.totalTreeSum(root)
    self.sumUnder(root)
    return int(self.res % (1e9 + 7))

  def totalTreeSum(self, root: TreeNode) -> None:
    if not root:
      return 0

    self.total += root.val
    self.totalTreeSum(root.left) and self.totalTreeSum(root.right)

  def sumUnder(self, root: TreeNode) -> int:
    if not root:
      return 0

    sumUnderLeft, sumUnderRight = self.sumUnder(
        root.left), self.sumUnder(root.right)
    self.res = max(self.res, (self.total - sumUnderLeft) *
                   sumUnderLeft, (self.total - sumUnderRight) * sumUnderRight)

    return sumUnderRight + sumUnderLeft + root.val

File: leetcode/medium/1339-max-prod-splitted/two-passes.cpp
/**
 * @file two-passes.cpp
 * @author Vortexx2
 * @brief Problem 1339 - Max Prod of splitted BT
 * @date 21-08-2021
 * 
 * Runtime - 116 ms
 * Memory Usage - 90.6 MB
 */
#include <algorithm>
#include <climits>
#include <deque>
#include <iostream>
#include <list>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  long long res = 0;
  int maxProduct(TreeNode *root) {
    long long total = totalTreeSum(root);
    sumUnder(root, total);
    return res % (int)(1e9 + 7);
  }

  long long totalTreeSum(TreeNode *root) {
    if (!root) return 0;
    return root->val + totalTreeSum(root->left) + totalTreeSum(root->right);
  }

  int sumUnder(TreeNode *root, long long &total) {
    if (!root) return 0;

    int sumUnderLeft = sumUnder(root->left, total),
        sumUnderRight = sumUnder(root->right, total);

    res = max({res, (total - sumUnderLeft) * sumUnderLeft,
               (total - sumUnderRight) * sumUnderRight});
    return sumUnderLeft + sumUnderRight + root->val;
  }
};